### PR TITLE
Fix to #1747 - Multi-select with Contains() does not generate SQL

### PR DESCRIPTION
**Pull request by maumar**
Fix is to translate Where(c => [list_of_constants].Constains(c.SomeProperty)) to
c.SomeProperty IN ( [list_of_constants)

We also translate !Contains... into NOT IN. Relinq converts Contains call into subquery, so we are detecting this pattern in the FilteringExpressionTreeVisitor and translate it accordingly. Argument to contains in this case will always be in the form of parameter. Parameter values are wired in during SQL generation.
Refactored EqualityPredicateOptimizer to factor in In/NotIn nodes created via Contains in the optimizations.

### CODE REVIEW

**Comment by maumar** [comment]
@anpete

**Comment by bricelam** [comment]
ProTip: You can update a PR by pushing more commits to the branch. See [Understanding the GitHub Flow](https://guides.github.com/introduction/flow/index.html).

**Comment by maumar** [comment]
thanks @bricelam! Updated this pull request with latest changes. @anpete

File: src/EntityFramework.Relational/Query/Sql/DefaultSqlQueryGenerator.cs [hunk 1]
```diff
@@ -259,24 +260,70 @@ public virtual Expression VisitInExpression(InExpression inExpression)
 
             _sql.Append(" IN (");
 
-            VisitJoin(inExpression.Values);
```
Comment (anpete) [inline src/EntityFramework.Relational/Query/Sql/DefaultSqlQueryGenerator.cs]: Why can't you just visit join the values.

File: src/EntityFramework.Relational/Query/Sql/DefaultSqlQueryGenerator.cs [hunk 2]
```diff
@@ -259,24 +260,70 @@ public virtual Expression VisitInExpression(InExpression inExpression)
...
+                    VisitConstantExpression(inConstant);
+                }
+
+                var inParameter = inValue as ParameterExpression;
+                if (inParameter != null)
+                {
+                    var parameterValue = _parameterValues[((ParameterExpression)inExpressionBase.Values.First()).Name];
+                    var valuesCollection = parameterValue as IEnumerable;
+
+                    if (valuesCollection != null && parameterValue.GetType() != typeof(string))
```
Comment (anpete) [inline src/EntityFramework.Relational/Query/Sql/DefaultSqlQueryGenerator.cs]: byte[]?

File: src/EntityFramework.Relational/Query/Sql/DefaultSqlQueryGenerator.cs [hunk 3]
```diff
@@ -259,24 +260,70 @@ public virtual Expression VisitInExpression(InExpression inExpression)
...
-            VisitJoin(inExpression.Values);
+            ProcessInExpressionArguments(notInExpression);
 
             _sql.Append(")");
 
-            return inExpression;
+            return notInExpression;
+        }
+
+        private void ProcessInExpressionArguments(InExpressionBase inExpressionBase)
```
Comment (anpete) [inline src/EntityFramework.Relational/Query/Sql/DefaultSqlQueryGenerator.cs]: Call this VisitInExpressionValues and make it protected virtual.

File: src/EntityFramework.Relational/Query/Sql/DefaultSqlQueryGenerator.cs [hunk 4]
```diff
@@ -259,24 +260,70 @@ public virtual Expression VisitInExpression(InExpression inExpression)
...
+                var inConstant = inValue as ConstantExpression;
+                if (inConstant != null)
+                {
+                    VisitConstantExpression(inConstant);
+                }
+
+                var inParameter = inValue as ParameterExpression;
+                if (inParameter != null)
+                {
+                    var parameterValue = _parameterValues[((ParameterExpression)inExpressionBase.Values.First()).Name];
```
Comment (anpete) [inline src/EntityFramework.Relational/Query/Sql/DefaultSqlQueryGenerator.cs]: Is it correct to call First() here?

File: src/EntityFramework.Relational/Query/ExpressionTreeVisitors/SqlTranslatingExpressionTreeVisitor.cs [hunk 5]
```diff
@@ -215,13 +217,12 @@ protected override Expression VisitUnaryExpression([NotNull] UnaryExpression exp
                 case ExpressionType.Not:
                 {
                     var operand = VisitExpression(expression.Operand);
+                    var inExpression = operand as InExpression;
 
-                    if (operand != null)
-                    {
-                        return Expression.Not(operand);
-                    }
+                    return inExpression != null
```
Comment (anpete) [inline src/EntityFramework.Relational/Query/ExpressionTreeVisitors/SqlTranslatingExpressionTreeVisitor.cs]: I wonder if we should remove NotInExpression and just use Not(InExpression). We would then need to handle this in SQL gen.

File: src/EntityFramework.Relational/Query/Sql/DefaultSqlQueryGenerator.cs [hunk 6]
```diff
@@ -562,22 +609,32 @@ protected override Expression VisitUnaryExpression(UnaryExpression unaryExpressi
 
             if (unaryExpression.NodeType == ExpressionType.Not)
             {
-                _sql.Append("NOT ");
+                var inExpression = unaryExpression.Operand as InExpression;
+                if (inExpression != null)
+                {
+                    return VisitNotInExpression(inExpression);
+                }
+                else
```
Comment (maumar) [inline src/EntityFramework.Relational/Query/Sql/DefaultSqlQueryGenerator.cs]: will remove the else

File: src/EntityFramework.Relational/Query/Sql/DefaultSqlQueryGenerator.cs [hunk 7]
```diff
@@ -259,26 +260,72 @@ public virtual Expression VisitInExpression(InExpression inExpression)
...
+                    VisitConstantExpression(inConstant);
+                }
+
+                var inParameter = inValue as ParameterExpression;
+                if (inParameter != null)
+                {
+                    var parameterValue = _parameterValues[inParameter.Name];
+                    var valuesCollection = parameterValue as IEnumerable;
+
+                    if (valuesCollection != null && parameterValue.GetType() != typeof(string) && parameterValue.GetType() != typeof(byte[]))
```
Comment (anpete) [inline src/EntityFramework.Relational/Query/Sql/DefaultSqlQueryGenerator.cs]: Break this line.

**Comment by anpete** [comment]
:shipit: Good work.

**Pull request merged**